database/memory/db.go

```diff
@@ -642,15 +642,37 @@ func (db *MemoryDB) GetERC20Balance(contract types.Address, holder types.Address
 	balanceMap := make(map[uint64]*big.Int)
 	frmBlkNum := options.BeginBlockNumber.Uint64()
 	endBlkNum := options.EndBlockNumber.Int64()
+	var maxEntry ERC20TokenHolder
+	maxEntryFound := false
 	for _, b := range db.erc20BalancesDB {
-		if contract == b.Contract && holder == b.Holder && b.BlockNumber >= frmBlkNum && (b.BlockNumber <= uint64(endBlkNum) || endBlkNum == -1) {
-			tokAmt, success := new(big.Int).SetString(b.Amount, 10)
-			if !success {
-				return nil, errors.New("could not parse token value")
+		if contract == b.Contract && holder == b.Holder {
+			if b.BlockNumber >= frmBlkNum && (b.BlockNumber <= uint64(endBlkNum) || endBlkNum == -1) {
+				tokAmt, success := new(big.Int).SetString(b.Amount, 10)
+				if !success {
+					return nil, errors.New("could not parse token value")
+				}
+				balanceMap[b.BlockNumber] = tokAmt
+			} else if len(balanceMap) == 0 {
+				if !maxEntryFound {
+					maxEntry = b
+					maxEntryFound = true
+				} else {
+					if maxEntry.BlockNumber < b.BlockNumber {
+						maxEntry = b
+					}
+				}
 			}
-			balanceMap[b.BlockNumber] = tokAmt
 		}
 	}
+
+	if len(balanceMap) == 0 && maxEntryFound {
```

Comment:
```suggestion
	if _, ok := balanceMap[fromBlk]; !ok && maxEntryFound {
```

We want to add an entry for the starting point only if one doesn't exist already. There are likely to be other entries in the balance map already for the blocks between the start and end, so checking if the map is empty won't work.

Reply:
agreed, i have fixed it.